src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+                {
+                    typeNode = (TTypeNode)generator.TypeExpression(
+                        indexerType ?? semanticModel.Compilation.GetSpecialType(SpecialType.System_Object));
+                }
+            }
+
+            // If we couldn't find an appropriate existing variable to use as the foreach
+            // variable, then generate one automatically.
+            if (foreachIdentifier.RawKind == 0)
+            {
+                foreachIdentifier = semanticFacts.GenerateUniqueName(
+                    semanticModel, forStatement, containerOpt: null, baseName: "v", cancellationToken);
+                foreachIdentifier = foreachIdentifier.WithAdditionalAnnotations(RenameAnnotation.Create());
+            }
+
+            // Create the expression we'll use to replace all matches in the for-body.
+            var foreachIdentifierReference = foreachIdentifier.WithoutAnnotations(RenameAnnotation.Kind).WithoutTrivia();
+
+            // Walk the for statement, replacing any matches we find.
+            findAndReplaceMatches(forStatement);
+
+            // Finally, remove the declaration statement if we found one.  Move all its leading
+            // trivia to the next statement.
+            if (declarationStatement != null)
+            {
+                editor.RemoveNode(declarationStatement,
+                    SyntaxGenerator.DefaultRemoveOptions | SyntaxRemoveOptions.KeepLeadingTrivia);
+            }
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(
+                forStatement,
+                (currentFor, _) => this.ConvertForNode(
+                    (TForStatementSyntax)currentFor, typeNode, foreachIdentifier,
+                    collectionExpression, iterationType, options));
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // local functions
+            (TTypeNode, SyntaxToken, TStatementSyntax) tryDeconstructInitialDeclaration()
+            {
+                if (bodyStatements.Count >= 1)
+                {
+                    var firstStatement = bodyStatements[0];
+                    if (syntaxFacts.IsLocalDeclarationStatement(firstStatement))
+                    {
+                        var variables = syntaxFacts.GetVariablesOfLocalDeclarationStatement(firstStatement);
+                        if (variables.Count == 1)
+                        {
+                            var firstVariable = (TVariableDeclaratorSyntax)variables[0];
+                            if (IsValidVariableDeclarator(firstVariable))
+                            {
+                                var firstVariableInitializer = syntaxFacts.GetValueOfEqualsValueClause(
+                                    syntaxFacts.GetInitializerOfVariableDeclarator(firstVariable));
+                                if (syntaxFacts.AreEquivalent(firstVariableInitializer, indexExpression))
```

Comment:
this is nice. didn't know we have this pattern matching.

Reply:
ah, it actually just use compiler's  SyntaxFactory.AreEquivalent

didn't thought using that as pattern matching.

Reply:
That was the reason i wrote SyntaxFactory.AreEquivalent in the first place :)   It was originally written for "introduce local for ALL occurrences".  We want to find occurrences of the same code, regardless of whitespace/comments.

Reply:
ah, didn't know. I thought that SyntaxFactory.AreEquivalent was to detect top level and syntax changes detection. for versioning and etc

Reply:
That's IsEquivalentTo.  Note: it's unfortunate that there are two names that are so similar here.